### PR TITLE
Improve sync

### PR DESCRIPTION
**Pull request by dgageot**
+ Add more logs and fix some of them
+ Improve errors
+ Pass context to commands

### CODE REVIEW

File: pkg/skaffold/runner/runner.go [hunk 1]
```diff
@@ -271,12 +271,12 @@ func (r *SkaffoldRunner) Dev(ctx context.Context, out io.Writer, artifacts []*la
 			return ErrorConfigurationChanged
 		case len(changed.needsResync) > 0:
 			for _, s := range changed.needsResync {
-				if err := r.Syncer.Sync(s); err != nil {
+				color.Default.Fprintf(out, "Syncing %d files for %s\n", len(s.Copy)+len(s.Delete), s.Image)
+
+				if err := r.Syncer.Sync(ctx, s); err != nil {
 					logrus.Warnln("Skipping build and deploy due to sync error:", err)
 					return nil
```
Comment (r2d4) [inline pkg/skaffold/runner/runner.go]: I forgot to add this but if we get an error here, we should add that dirty artifact back to the rebuild list

File: pkg/skaffold/runner/runner.go [hunk 2]
```diff
@@ -271,12 +271,12 @@ func (r *SkaffoldRunner) Dev(ctx context.Context, out io.Writer, artifacts []*la
 			return ErrorConfigurationChanged
 		case len(changed.needsResync) > 0:
 			for _, s := range changed.needsResync {
-				if err := r.Syncer.Sync(s); err != nil {
+				color.Default.Fprintf(out, "Syncing %d files for %s\n", len(s.Copy)+len(s.Delete), s.Image)
```
Comment (r2d4) [inline pkg/skaffold/runner/runner.go]: update to latest code

File: pkg/skaffold/kubernetes/sync.go [hunk 3]
```diff
@@ -34,60 +34,76 @@ import (
...
+func perform(ctx context.Context, image string, files map[string]string, cmdFn func(context.Context, v1.Pod, v1.Container, string, string) *exec.Cmd) error {
 	client, err := Client()
 	if err != nil {
 		return errors.Wrap(err, "getting k8s client")
 	}
-	pods, err := client.CoreV1().Pods("").List(meta_v1.ListOptions{
-		LabelSelector: labelSelector(),
-	})
+
+	pods, err := client.CoreV1().Pods("").List(meta_v1.ListOptions{})
```
Comment (r2d4) [inline pkg/skaffold/kubernetes/sync.go]: use the skaffold labelselector here

File: pkg/skaffold/kubernetes/sync.go [hunk 4]
```diff
@@ -34,60 +34,76 @@ import (
...
+				})
 			}
 		}
 	}
+
+	if err := e.Wait(); err != nil {
+		return errors.Wrap(err, "syncing files")
+	}
+
+	if int(performed) != len(files) {
```
Comment (r2d4) [inline pkg/skaffold/kubernetes/sync.go]: I don't think you need to count here. errgroup handles this

File: pkg/skaffold/kubernetes/sync.go [hunk 5]
```diff
@@ -34,60 +34,76 @@ import (
 
 type KubectlSyncer struct{}
 
-func (k *KubectlSyncer) Sync(s *sync.Item) error {
-	if err := perform(s.Image, s.Copy, copyFileFn); err != nil {
-		return errors.Wrap(err, "copying files")
+func (k *KubectlSyncer) Sync(ctx context.Context, s *sync.Item) error {
+	if len(s.Copy) > 0 {
```
Comment (r2d4) [inline pkg/skaffold/kubernetes/sync.go]: maybe move the len check into `perform`? a few lines of code saved

File: pkg/skaffold/kubernetes/sync.go (hunk 3 above)
Comment (r2d4) [reply pkg/skaffold/kubernetes/sync.go]: ah good point 👍

File: pkg/skaffold/runner/runner.go (hunk 2 above)
Comment (r2d4) [reply pkg/skaffold/runner/runner.go]: Yeah just maybe the last commit. I changed this message to the debug and info statement.

File: pkg/skaffold/kubernetes/sync.go (hunk 4 above)
Comment (r2d4) [reply pkg/skaffold/kubernetes/sync.go]: Hm, I guess this is a bit tricky. What if we have 2 replicas?

File: pkg/skaffold/runner/runner.go (hunk 2 above)
Comment (dgageot) [reply pkg/skaffold/runner/runner.go]: Oh, the one I asked you to add? It didn't work well in reality :-) Didn't give enough feedback

Comment (r2d4) [reply pkg/skaffold/runner/runner.go]: :) SGTM

File: pkg/skaffold/kubernetes/sync.go (hunk 4 above)
Comment (dgageot) [reply pkg/skaffold/kubernetes/sync.go]: Good point, I'll see how to fix it. Each file should be synced at least once, I guess.

Comment (r2d4) [reply pkg/skaffold/kubernetes/sync.go]: Yeah I think thats the best we can do. But making sure we sync at least once is a good check.

File: pkg/skaffold/runner/runner.go (hunk 1 above)
Comment (dgageot) [reply pkg/skaffold/runner/runner.go]: How would you do that?

**Pull request merged**